### PR TITLE
Add autocomplete select component

### PR DESCRIPTION
**Pull request by JamieCleare2525**
## Context

- For ease of coding and reuse. An AutocompleteSelectFormComponent has replaced the repetitive code.

## Changes proposed in this pull request

- Implement AutocompleteSelectFormComponent
- Replace existing duplicated code to use AutocompleteSelectFormComponent

## Guidance to review

_(Claims)_
- Sign in as Colin
- Test Onboarding Schools

_(Placements)_
- Sign in as Colin
- Test Onboarding Schools
- Test Onboarding Providers

_Previews available at_
http://placements.localhost:3000/rails/view_components
http://placements.localhost:3000/rails/view_components/autocomplete_select_form_component/provider_onboarding_form
http://placements.localhost:3000/rails/view_components/autocomplete_select_form_component/school_onboarding_form

http://claims.localhost:3000/rails/view_components
http://claims.localhost:3000/rails/view_components/autocomplete_select_form_component/provider_onboarding_form
http://claims.localhost:3000/rails/view_components/autocomplete_select_form_component/school_onboarding_form

## Link to Trello card

https://trello.com/c/8EINwpQR/231-implement-a-viewcomponent-for-autocomplete-select-forms

## Screenshots

https://github.com/DFE-Digital/itt-mentor-services/assets/17162488/e7d8b5df-162e-42ae-a8a1-d27a2740a483

https://github.com/DFE-Digital/itt-mentor-services/assets/17162488/3b60fe12-31f5-49c6-ac54-e439b2234fd9

### CODE REVIEW

**Review by Nitemaeric — COMMENTED** [review]
Hmm, before I look further into this PR.

Would it make more sense for us to make the input with autocomplete functionality into a component and leave the rest of this current component's functionality outside of the component? Such as, the back button, error summary, etc.

**Comment by JamieCleare2525** [comment]
> Hmm, before I look further into this PR.
> 
> Would it make more sense for us to make the input with autocomplete functionality into a component and leave the rest of this current component's functionality outside of the component? Such as, the back button, error summary, etc.

This component follows the same format as the `SchoolOptionsFormComponent` and `ProviderOptionsFormComponent` @CatalinVoineag made 3 months ago.

But, due to the variable changes to the `data` attributes which are required as part of the form, and also reflect changes in the form itself (such as the `data-input-name` on the div which becomes the  displayed input containing the `name` of the selected dropdown item). This component is more reusable as a complete form. This also would therefore require the `error_summary` to be included as part of the component.

File: app/components/autocomplete_select_form_component.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+class AutocompleteSelectFormComponent < ApplicationComponent
+  attr_reader :model, :scope, :url, :data, :input
+
+  # data: { :turbo, :controller, :autocomplete_path_value,
```
Comment (CatalinVoineag) [inline app/components/autocomplete_select_form_component.rb:4]: Are these meant to be here?

**Review by CatalinVoineag — APPROVED** [review]
Overall this looks good to me, well done!

File: app/components/autocomplete_select_form_component.rb (hunk 1 above)
Comment (JamieCleare2525) [reply app/components/autocomplete_select_form_component.rb:4]: I figured they would be good to keep in, as useful information for developers using the component.

File: app/views/claims/support/schools/new.html.erb [hunk 2]
```diff
@@ -6,41 +6,23 @@
...
-                                              previous_search: @school_form&.school&.id } %>
-
-          <div class="govuk-!-margin-bottom-7" data-autocomplete-target="input" data-input-name="school[name]"
-            data-action="focusin->autocomplete#clearUndefinedSuggestions click->autocomplete#clearUndefinedSuggestions">
-          </div>
-        <% end %>
-
-        <%= f.govuk_submit t(".continue") %>
+    input: {
+      value: @school_form&.school&.name,
```
Comment (Kizr) [inline app/views/claims/support/schools/new.html.erb:18]: Could we delegate this?

**Review by ollietreend — APPROVED** [review]
This looks good to me 👍🏻 (pending the existing comments)

File: app/views/claims/support/schools/new.html.erb [hunk 3]
```diff
@@ -6,41 +6,23 @@
...
+      label: t(".title"),
+      caption: t(".caption"),
+      previous_search: @school_form&.school&.id,
```
[the remark this answers is not in the document]
Comment (JamieCleare2525) [reply app/views/claims/support/schools/new.html.erb]: resolved in [3d12c4e](https://github.com/DFE-Digital/itt-mentor-services/pull/459/commits/3d12c4e56e64c7ce783dcedf608664d4901025d2)

File: app/views/claims/support/schools/new.html.erb (hunk 2 above)
Comment (JamieCleare2525) [reply app/views/claims/support/schools/new.html.erb]: resolved in [3d12c4e](https://github.com/DFE-Digital/itt-mentor-services/pull/459/commits/3d12c4e56e64c7ce783dcedf608664d4901025d2)

**Review by Kizr — APPROVED** [review]

**Pull request merged**